students/rhamersky/Lesson_4/mailroom_part2.py

```diff
@@ -0,0 +1,335 @@
+            if len(tup_donor_names) > 0:
+                donor_list_creation()  # --> Creates a list of donors
+                bool_flag = False  # --> Is used as a flag on the while loop below
+            else:
+                print("\n"+"There are no donors in list.")  # --> Let's the user know that a donor list does not exist.
+                bool_flag = False  # --> Is used as a flag on the while loop below.
+            break
+
+        # Gathers donation information from the user and prints a thank-you message for the donor.
+        while bool_flag:
+            try:
+                flo_donation_amount = float(input(str_name + ", please enter donation amount. "))
+
+                # Updates donor tuple
+                update_donor(donor_name, flo_donation_amount)
+
+                # String formatting for a brief thank-you to donor.
+                thankyou_message = "\n"+"{}, thank you for donating ${:,.2f}.".format(donor_name, flo_donation_amount)
+                print(thankyou_message)
+
+            except ValueError:  # --> Allows the program to continue if the user does not enter the right data type.
+                print(str_name + ", not a valid amount. \n")
+                continue
+
+            else:
+                break
+
+    @staticmethod
+    def create_report():
+        '''
+        This method takes the table tuple and creates a summary report for the user (name, number of gifts, total amount
+        in gifts, and the average amount given. The code sorts the list rows into the largest amount amount give to the
+        least amount given.
+        :return:
+        '''
+        total = int()
+        donation_count = int()
+        lst_table = []
+        # Creates a new list [name, number of gifts, average amount given, total amount in gifts] to be used to create
+        # the summary chart.
+        for row in tup_donor_names:
+            for value in range(2, len(row)):
+                total += row[value]
+                donation_count += 1
+            lst_table.append([row[0], row[1], total/donation_count, total])
+            total = 0
+            donation_count = 0
+        # Reverses the list
+        for row in lst_table:
+            row.reverse()
+        # Shorts the list form largest to smallest on the first element of each row.
+        lst_table.sort(reverse=True)
+
+        index = 0
+        # Formatting the money elements and changing all int() & float() data to str() data
+        # to use for creating the chart.
+        for row in range(len(lst_table)):
+            lst_table[row][index] = "${:,.2f}".format(lst_table[row][index])  # --> Total amount for gifts
+            lst_table[row][index + 1] = "${:,.2f}".format(lst_table[row][index + 1])  # --> Avg. gift amount
+            lst_table[row][index + 2] = str(lst_table[row][index + 2])  # --> Number of gifts
+        # Reverses list
+        for row in lst_table:
+            row.reverse()
```

Comment:
are above 4 loops necessary? can this be done in a single loop?

Reply:
I am not sure how to condense this section of code.

Reply:
this should replace 99 through 144:
```
        summary = []
        # {"Jeff Bezos": [100, 200]}
        for donor, data in donors.items():
            # name, total, number_of_donations, avg
            summary.append(
                donor, sum(data), len(data), sum(data) / len(data)
            )
            # you can find your max value in this loop as well

        # do sorting
        sorted_summary = sorted_operation_here # this is just a single line using sorted()

        # print heading

        for row in sorted_summary:
            # format your row
            print("".format(*row))
```

you now only have 3 loops and a cleaner flow to follow